src/ott/neural/flows/genot.py

```diff
@@ -0,0 +1,490 @@
+    self.flow = flow
+    self.time_sampler = time_sampler
+    self.optimizer = optimizer
+    self.checkpoint_manager = checkpoint_manager
+    self.latent_noise_fn = jax.tree_util.Partial(
+        jax.random.multivariate_normal,
+        mean=jnp.zeros((output_dim,)),
+        cov=jnp.diag(jnp.ones((output_dim,)))
+    )
+    self.input_dim = input_dim
+    self.output_dim = output_dim
+    self.cond_dim = cond_dim
+    self.k_samples_per_x = k_samples_per_x
+
+    # OT data-data matching parameters
+    self.ot_solver = ot_solver
+    self.epsilon = epsilon
+    self.cost_fn = cost_fn
+    self.scale_cost = scale_cost
+    self.fused_penalty = fused_penalty
+
+    # OT latent-data matching parameters
+    self.solver_latent_to_data = solver_latent_to_data
+    self.kwargs_solver_latent_to_data = kwargs_solver_latent_to_data
+
+    # callback parameteres
+    self.callback_fn = callback_fn
+    self.setup()
+
+  def setup(self):
+    """Set up the model.
+
+    Parameters
+    ----------
+    kwargs
+    Keyword arguments for the setup function.
+    """
+    self.state_velocity_field = (
+        self.velocity_field.create_train_state(
+            self.rng, self.optimizer, self.output_dim
+        )
+    )
+    self.step_fn = self._get_step_fn()
+    if self.solver_latent_to_data is not None:
+      self.match_latent_to_data_fn = self._get_sinkhorn_match_fn(
+          ot_solver=self.solver_latent_to_data,
+          **self.kwargs_solver_latent_to_data
+      )
+    else:
+      self.match_latent_to_data_fn = lambda key, x, y, **_: (x, y)
+
+    # TODO: add graph construction function
+    if isinstance(self.ot_solver, sinkhorn.Sinkhorn):
+      self.match_fn = self._get_sinkhorn_match_fn(
+          ot_solver=self.ot_solver,
+          epsilon=self.epsilon,
+          cost_fn=self.cost_fn,
+          scale_cost=self.scale_cost,
+          tau_a=self.tau_a,
+          tau_b=self.tau_b,
+          filter_input=True
+      )
+    else:
+      self.match_fn = self._get_gromov_match_fn(
+          ot_solver=self.ot_solver,
+          cost_fn=self.cost_fn,
+          scale_cost=self.scale_cost,
+          tau_a=self.tau_a,
+          tau_b=self.tau_b,
+          fused_penalty=self.fused_penalty
+      )
+
+  def __call__(self, train_loader, valid_loader):
+    """Train GENOT.
+
+    Args:
+      train_loader: Data loader for the training data.
+      valid_loader: Data loader for the validation data.
+    """
+    batch: Dict[str, jnp.array] = {}
+    for iteration in range(self.iterations):
+      batch = next(train_loader)
+
+      (
+          self.rng, rng_time, rng_resample, rng_noise, rng_latent_data_match,
+          rng_step_fn
+      ) = jax.random.split(self.rng, 6)
+      batch_size = len(
+          batch["source_lin"]
+      ) if batch["source_lin"] is not None else len(batch["source_quad"])
+      n_samples = batch_size * self.k_samples_per_x
+      batch["time"] = self.time_sampler(rng_time, n_samples)
```

Comment:
Since you always pass the arrays to the downstream functions, I don't think it's necessary to assign the newly created variables (or updated ones) to the `batch`.